method/CellCharter/CellCharter.py

```diff
@@ -0,0 +1,195 @@
+with open(args.config, "r") as f:
+    config = json.load(f)
+
+# Set up output files
+from pathlib import Path
+out_dir = Path(args.out_dir)
+label_file = out_dir / "domains.tsv"
+embedding_file = out_dir / "embedding.tsv"
+
+################# Define functions #################
+def get_anndata(args):
+    import anndata as ad
+    import numpy as np
+    import pandas as pd
+    import scipy as sp
+    from PIL import Image
+
+    observations = pd.read_table(args.observations, index_col=0)
+    features = pd.read_table(args.features, index_col=0)
+
+    coordinates = (
+        pd.read_table(args.coordinates, index_col=0)
+        .loc[observations.index, :]
+        .to_numpy()
+    )
+
+    adata = ad.AnnData(obs=observations, 
+                       var=features, 
+                       obsm={"spatial": coordinates})
+
+    if args.matrix is not None:
+        X = sp.io.mmread(args.matrix)
+        if sp.sparse.issparse(X):
+            X = X.tocsr()
+        adata.X = X
+
+    if args.neighbors is not None:
+        adata.obsp["spatial_connectivities"] = sp.io.mmread(args.neighbors).T.tocsr()
+
+    # Filter by selected samples/features
+    if "selected" in adata.obs.columns:
+        adata = adata[observations["selected"].astype(bool), :]
+    if "selected" in adata.var.columns:
+        adata = adata[:, features["selected"].astype(bool)]
+
+    if args.dim_red is not None:
+        adata.obsm["reduced_dimensions"] = (
+            pd.read_table(args.dim_red, index_col=0).loc[adata.obs_names].to_numpy()
+        )
+
+    if args.image is not None:
+        adata.uns["image"] = np.array(Image.open(args.image))
+    else:
+        adata.uns["image"] = None
+
+    return adata
+
+## Your code goes here
+import cellcharter as cc
+import anndata as ad
+import scanpy as sc
+import pandas as pd
+import numpy as np
+from squidpy._constants._pkg_constants import Key
+from sklearn.metrics.pairwise import euclidean_distances
+from scipy.sparse import csr_matrix
+from itertools import chain
+import random
+
+random.seed(seed)
+
+# Load data
+adata = get_anndata(args)
+adata.var_names_make_unique()
+
+# Generate spatial neighbor distance (from Squidpy source code)
+coords = adata.obsm["spatial"]
+N = coords.shape[0]
+Adj = adata.obsp["spatial_connectivities"] 
+
+# Calculate euclidean distances
+dists = np.array(list(chain(*(
+    euclidean_distances(coords[Adj.indices[Adj.indptr[i] : Adj.indptr[i + 1]], :], 
+                        coords[np.newaxis, i, :])
+    for i in range(N)
+    if len(Adj.indices[Adj.indptr[i] : Adj.indptr[i + 1]])
+)))).squeeze()
+
+Dst = csr_matrix((dists, Adj.indices, Adj.indptr), shape=(N, N))
+Dst.setdiag(0.0)
+
+# Fitting into the respective data slot
+adata.obsp["spatial_distances"] = Dst
+adata.uns["spatial_neighbors"] = {
+        "connectivities_key": "spatial_connectivities",
+        "distances_key": "spatial_distances",
+        "params": {"coord_type": "generic", "delaunay": True, "transform": None},
```

Comment:
These params might not always be correct. Is this used in later steps of the algorithm?

Reply:
Hi, these parameters aren't really used in any of the downstream analyses. I simply put in there as the code for calculating the spatial distance is taken from the case where `delaunay=True` and `coord_type=generic` in [`squidpy.gr.spatial_neighbors`](https://squidpy.readthedocs.io/en/stable/api/squidpy.gr.spatial_neighbors.html). 

The main point of code lines 143-159 is to replicate how the squidpy function generates the distance matrix from the adjacency matrix. And here is what each parameter is for: 
* `delaunay=True`: It is set to be `True` because in cellcharter tutorial they mentioned the delaunay parameters need to be true for downstream functions. I also adapted the code for the distance matrix assuming `delaunay=True` in the function.
* `coord_type=generic`: If `delaunay=True`, this parameter doesn't make a difference, set to default
* `transform=None`: Set to default, assuming no transformation on the adjacency matrix.

Hope it answers the question! Those parameters don't affect the subsequent analysis but the dictionary structure needs to be there. Let me know how you think we could change it! :))

Reply:
So do they always assume delaunay triangulation as input for the neighbor structure? Because they also do this n-th percentile trimming of the neighbor graph which does not necessarily make sense for some other methods, e.g. anything that is based of neighbors on a regular grid which is usually used for Visium for example.

Reply:
Ohh that's a good point! Sorry I missed that earlier. In their documentation, they only showcase the example of CosMx and they use `delaunay=True`. But in their [paper](https://www.nature.com/articles/s41588-023-01588-4#Sec7) they did mention that removing long links is unnecessary for regular structure technology. 
I will make it optional based on the `technology` option. If `technology==Visium`, I will skip the distance calculation and the `remove_long_links` part, as the distance matrix is only needed for`remove_long_links`.

Reply:
It should also be done for "ST" not only "Visium".

The issue I see that it is technically not only dependent on the technology but also on the neighbor definition. e.g. you could define neighbors based on distance and then doing a trimming on the 99th percentile does not make much sense.

Not quite sure what the best approach is for now.

Reply:
Since the tirmming only makes sense when delaunay triangulation is applied to construct the neighborhood graph (hence not depending on the distance), I can try to implement some codes to check if the neighbor graphs provided fulfills delaunay criteria and proceed on trimming based on that. Do you think that'd be a way to work around it? :))